File: reference/node/src/render.ts
/**
 * Renderers turn arc.status events into human-facing strings for a given
 * surface (terminal, Slack, Telegram, etc.).
 *
 * Rendering is intentionally a consumer concern (see spec §1.3 design
 * principle 5: "Receiver renders"). These helpers are provided so common
 * cases work in one line; complex surfaces should write their own.
 */

import type { ArcStatusEvent, ArcStatusPhase } from "./types.js";

const PHASE_SYMBOLS: Record<ArcStatusPhase, string> = {
  started: "▶",
  milestone: "✓",
  heartbeat: "·",
  done: "■",
  blocked: "⛔",
};

const PHASE_LABELS: Record<ArcStatusPhase, string> = {
  started: "STARTED",
  milestone: "MILESTONE",
  heartbeat: "HEARTBEAT",
  done: "DONE",
  blocked: "BLOCKED",
};

export interface RenderOptions {
  /** Include a leading symbol per phase. Default: true. */
  symbol?: boolean;
  /** Include the phase label in CAPS. Default: false. */
  phaseLabel?: boolean;
  /** Include `step/total` if both are present on the event. Default: true. */
  step?: boolean;
  /** Include `eta_minutes` if present. Default: true. */
  eta?: boolean;
  /** Append the body, if present. Default: false. */
  body?: boolean;
  /** Truncate the body to N characters when included. Default: 240. */
  bodyMax?: number;
}

const DEFAULTS: Required<RenderOptions> = {
  symbol: true,
  phaseLabel: false,
  step: true,
  eta: true,
  body: false,
  bodyMax: 240,
};

/**
 * Render an event to a single human-readable line, with optional body block.
 *
 * Examples:
 *   "▶ build Pulsefeed v0.1"
 *   "✓ [6/11] receiver booted (ETA 25m)"
 *   "· still working: hyperparameter sweep (run 4/12)"
 *   "■ v0.1 complete, 43 tests, deployed"
 *   "⛔ need finance sign-off on plan-B reclassification"
 */
export function render(event: ArcStatusEvent, options?: RenderOptions): string {
  const opts = { ...DEFAULTS, ...options };

  const parts: string[] = [];

  if (opts.symbol) {
    parts.push(PHASE_SYMBOLS[event.phase]);
  }

  if (opts.phaseLabel) {
    parts.push(PHASE_LABELS[event.phase]);
  }

  if (
    opts.step &&
    typeof event.step === "number" &&
    typeof event.total === "number"
  ) {
    parts.push(`[${event.step}/${event.total}]`);
  }

  parts.push(event.title);

  if (opts.eta && typeof event.eta_minutes === "number") {
    parts.push(`(ETA ${formatMinutes(event.eta_minutes)})`);
  }

  let line = parts.join(" ");

  if (opts.body && event.body) {
    const max = Math.max(1, opts.bodyMax);
    const body =
      event.body.length > max
        ? event.body.slice(0, max - 1) + "…"
        : event.body;
    line += "\n" + body;
  }

  return line;
}

function formatMinutes(minutes: number): string {
  if (minutes < 1) return "<1m";
  // Round to whole minutes first, then split into h/m, so a value like 119.6
  // becomes 2h rather than the malformed "1h60m".
  const total = Math.round(minutes);
  if (total < 60) return `${total}m`;
  const h = Math.floor(total / 60);
  const m = total - h * 60;
  if (m === 0) return `${h}h`;
  return `${h}h${m}m`;
}

File: reference/python/src/agent_arc_status/render.py
"""Default human-line renderer. Rendering is a consumer concern (spec §1.3);
these helpers cover common cases in one call."""

from __future__ import annotations

import math
from typing import cast

from .types import ArcStatusEvent, ArcStatusPhase

_PHASE_SYMBOLS: dict[ArcStatusPhase, str] = {
    "started": "▶",
    "milestone": "✓",
    "heartbeat": "·",
    "done": "■",
    "blocked": "⛔",
}

_PHASE_LABELS: dict[ArcStatusPhase, str] = {
    "started": "STARTED",
    "milestone": "MILESTONE",
    "heartbeat": "HEARTBEAT",
    "done": "DONE",
    "blocked": "BLOCKED",
}


def _is_int(value: object) -> bool:
    return isinstance(value, int) and not isinstance(value, bool)


def _is_number(value: object) -> bool:
    return isinstance(value, (int, float)) and not isinstance(value, bool)


def render(
    event: ArcStatusEvent,
    *,
    symbol: bool = True,
    phase_label: bool = False,
    step: bool = True,
    eta: bool = True,
    body: bool = False,
    body_max: int = 240,
) -> str:
    """Render an event to a single line, with an optional appended body block.

    Examples::

        "▶ build atlas index"
        "✓ [6/11] receiver booted (ETA 25m)"
        "■ v0.1 complete, 43 tests, deployed"
    """
    parts: list[str] = []

    if symbol:
        parts.append(_PHASE_SYMBOLS[event["phase"]])
    if phase_label:
        parts.append(_PHASE_LABELS[event["phase"]])

    step_v = event.get("step")
    total_v = event.get("total")
    if step and _is_int(step_v) and _is_int(total_v):
        parts.append(f"[{step_v}/{total_v}]")

    parts.append(event["title"])

    eta_v = event.get("eta_minutes")
    if eta and _is_number(eta_v):
        parts.append(f"(ETA {_format_minutes(cast(float, eta_v))})")

    line = " ".join(parts)

    body_v = event.get("body")
    if body and body_v:
        limit = max(1, body_max)
        rendered = body_v[: limit - 1] + "…" if len(body_v) > limit else body_v
        line += "\n" + rendered

    return line


def _format_minutes(minutes: float) -> str:
    if minutes < 1:
        return "<1m"
    # Round half-up to match JS Math.round (Python's round() is half-to-even),
    # then split into h/m so 119.6 -> "2h" rather than the malformed "1h60m".
    total = math.floor(minutes + 0.5)
    if total < 60:
        return f"{total}m"
    h = total // 60
    m = total - h * 60
    if m == 0:
        return f"{h}h"
    return f"{h}h{m}m"
